### ISSUE TITLE
minimal Class support

### ISSUE THREAD

**Issue by hohwille** [cincheo/jsweet#181]
In Java you have things like reflection with Type and Class. Just like GWT I understand the JSweet never wants to support reflection and that is for good reason (requires ultra complex and heavyweight JVM emulation). However, all this approach to write client code in Java is IMHO also about reuse of code on server AND client. If you as JSweet makes disagree on this please let me know.

In Java it is very common to have code doing things like getClass(). E.g. assume just to call getSimpleName() on the class object. In TypeScript you also have classes and internally there are prototypes so it should be possible to create a minimal shim to support basic things. E.g. for GWT I wrote an implementation of SLF4J that works in the browser and can redirect to `console.log` or collect the logs and send them in chunks to the server or whatever you want.
Would this possible as a JSweet Candy?
JSweet would have to be able to compile code like
```
private static final Logger LOG = LoggerFactory.getLogger(MyClass.class);
```

If I have expectations that contradict the JSweet philosophy then let me know. Otherwise if there could be a plan how to allow this (maybe even optionally if it would impact the footprint) please let me know. I would also actively support e.g. with PRs if I get some hints and spots where to get started.

**Comment by renaudpawlak** [comment]
What I consider as being reflection, is all that has to do with field and method manipulation. It is unlikely that JSweet support ``Class.getMethods()`` or ``Class.getDeclaredField(String name)`` for instance... 

Class and class name access on the other hand are necessary and are supported by JSweet (``*.class``, ``Object.getClass()``, ``Class.getName()`` and ``Class.getSimpleName()`` are supported in most cases). Here is how it works so far:

- ``X.class`` => is replaced by the class constructor référence when ``X`` is a class, with the fully qualified name string when ``X`` is an interface (remember that interfaces are not available at runtime with TypeScript). Also, the class constructor contains an additional field called ``__classname`` which contains the fully qualified name of the class and is used when accessing the name of the class.
- ``X.class.getName()``and ``X.class.getSimpleName()`` => are substituted by their actual values by JSweet.
- ``o.getClass()`` => returns the o's class constructor
- ``o.getClass().getName()`` and ``o.getClass().getSimpleName()`` => use the ``__classname`` field in the constructor to resolve the class names.

So, in conclusion, JSweet should supports the following code and it does not contradict its initial philosophy (you just have to understand that the class is replaced by the JavaScript constructor at runtime):

``private static final Logger LOG = LoggerFactory.getLogger(MyClass.class);``

**Comment by hohwille** [comment]
Awesome!!! 
But I tweaked your QuickStart.java class with the following code:
```
    String string = "";
    a.push(string.getClass().getName());
```

This results in the following error:
```
2016-11-04 09:20:23.023 INFO  JSweetTranspiler:1081 - quickstart/QuickStart.ts(17,27): error TS2339: Property 'getClass' does not exist on type 'string'.
ERROR: property 'getClass' does not exist on type 'string' at C:\Projekte\mmm\workspaces\test\jsweet-quickstart\src\main\java\quickstart\QuickStart.java(26)
2016-11-04 09:20:23.023 INFO  JSweetTranspiler:616 - transpilation process finished in 2849 ms
[ERROR] transpilation failed
org.apache.maven.plugin.MojoFailureException: transpilation failed with 1 error(s) and 3 warning(s)
```

**Comment by hohwille** [comment]
OK. It works with this:
```
    a.push(QuickStart.class.getName());
```
and gives expected output. Cool.
Shall I file a new bug that this does not work for internal classes such as String? Or can we reopen and rename this one?

**Comment by renaudpawlak** [comment]
Umm. I have tried your code and it worked for me. 

Maybe you have hit a bug because ``string`` is a type in TypeScript. JSweet may fail to substitute the ``getClass().getName()`` macro in that case (but it seems to work with me)... Can you try to rename the ``string`` variable with ``str``? Does it work better? If it still does not work, can you show the full code and the generated TypeScript?

**Comment by renaudpawlak** [comment]
It looks like you are compiling with Maven. I was using the Eclipse plugin. Which version of the Maven plugin are you using? Make sure you use the 1.2.0-SNAPSHOT to get the latest features.

**Comment by renaudpawlak** [comment]
Note: ``Object.getClass()`` was supported in 1.2.0-SNAPSHOT and was not supported in JSweet 1.1. That would explain it all. Could you confirm and close?

**Comment by hohwille** [comment]
Yes, I am using Maven as Eclipse plugin does not work at all for me (see https://github.com/cincheo/jsweet-eclipse-plugin/issues/12). All I did is clone your own example:
https://github.com/cincheo/jsweet-quickstart
And inserted the above quoted lines. Once again:
```
String string = "";
a.push(string.getClass().getName());
```
That is all. But you are right in your POM you are using 1.1.1 of the maven plugin:
https://github.com/cincheo/jsweet-quickstart/blob/master/pom.xml#L63
But JSweet Core 1.2.0-SNAPSHOT:
https://github.com/cincheo/jsweet-quickstart/blob/master/pom.xml#L123

I locally updated the maven plugin to `1.2.0-SNAPSHOT` and now it compiles. Great progress.
BUT: The result is wrong "".getClass().getName() returns the empty String instead of "String" or "java.lang.String". The question is if that is to be considered correct in the context of JSweet.
From a Java perspective this is wrong. But I can see that some compromises have to be made...
At least it is great that I can get a Class object and I will check if all Strings will return the same Class object instance while numbers or other objects return something else. Then at least it is possible to write a generic factory that uses the Class of an object as key.
At least to confirm:
* `class` literals are working
* `getClass()` is working with recent versions
* `Class.getName()` and `Class.getSimpleName()` does not work as expected for some internal types such as `java.lang.String` (what is technically a String in the browser where nothing such as `java.lang.String` actually exists).
Thanks for the hint with the recent version and thanks for all your work on JSweet. :)

**Comment by hohwille** [comment]
The behaviour is from JavaScript:
```
    var string = "";
    var sc = string.constructor;
    alert(sc["__classname"]);
```
Shows a popup with `undefined`.

**Comment by renaudpawlak** [comment]
Ok. I will do more tests. You probably cannot modify constructors of built-in JS types. So it may be a bug and we need a workaround for JS built-in types. (your case would probably work for a user class, right?). Thanks for reporting!

**Comment by hohwille** [comment]
Jep. So the initial bug report is kind of invalid (custom classes work as expected). However, build in JS types need a special handling. actually `"".constructor` results in an object that is a function string ...
So if I would need a workaround urgently I might be able to call a toString() on the class and strip the name somehow from there. But it would be awesome if you could provide a smart custom handling inside JSweet for these JS types.

**Comment by renaudpawlak** [comment]
I will. Just don't know when. I'll keep you informed.

**Comment by renaudpawlak** [comment]
Is it OK to return ``String``, ``Number`` and ``Boolean`` for class names when objects are primitive types? These correspond to the JavaScript primitive types constructor names. I don't think that they should be prefixed with ``java.lang.``. So, that would be:

```Java
"".getClass().getName(); // "String"
2.getClass().getName(); // "Number"
true.getClass().getName(); // "Boolean"
```

What do you think?

**Comment by renaudpawlak** [comment]
I have committed the proposal above so that you can play with it. Don't hesitate to comment on it.

**Comment by hohwille** [comment]
Tested with latest snapshot version and works as described.
For shared code that is used both on server and client it is IMHO always better to emulate Java methods the Java way (as there is no `getClass().getName()` in JS anyways). So actually the `java.lang.` prefix would be added then and only omitted for `getSimpleName()`.
However, I am used to writing code for such purpose in a way to workaround this. In GWT I did a lot of super-sourcing to workaround limitations...
In the end it is a matter of a design-goal if JSweet should be more of a Java emulator for Java specific stuff or whether it should only give the minimal support that makes sense for JS world but will allow calling Java APIs.
Thanks for fixing the actual issue. This is great progress :+1:

**Comment by renaudpawlak** [comment]
Thanks! About adding the ``java.lang`` prefix, that would indeed be easy. However, it would not work anyway for numbers (int, long, short, ...) because it would always return ``java.lang.Number``. That's why I did not do it. 

I believe that comparing class names with strings is bad practice, even in Java and should be avoided anyway, but on the other hand, if we can ensure a little more compatibility between the client and the server, it doesn't hurt :D

I reopen the issue to think more about it...

**Comment by renaudpawlak** [comment]
So far it seems to me that it is better keep it the way it is because we don't want JSweet to rely on Java conventions. We will see what we do if a real use case pops out. Thanks.
